Firestore/core/test/firebase/firestore/remote/grpc_connection_test.cc

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/auth/token.h"
+#include "Firestore/core/src/firebase/firestore/remote/connectivity_monitor.h"
+#include "Firestore/core/src/firebase/firestore/remote/grpc_connection.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/executor_std.h"
+#include "Firestore/core/test/firebase/firestore/util/grpc_stream_tester.h"
+#include "absl/memory/memory.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+using auth::Token;
+using auth::User;
+using core::DatabaseInfo;
+using model::DatabaseId;
+using util::AsyncQueue;
+using util::GrpcStreamTester;
+using util::Status;
+using util::StatusOr;
+using util::internal::ExecutorStd;
+using NetworkStatus = ConnectivityMonitor::NetworkStatus;
+
+namespace {
+
+class FakeConnectivityMonitor : public ConnectivityMonitor {
+ public:
+  explicit FakeConnectivityMonitor(AsyncQueue* worker_queue)
+      : ConnectivityMonitor{worker_queue} {
+    SetInitialStatus(NetworkStatus::Available);
+  }
+
+  void set_status(NetworkStatus new_status) {
+    MaybeInvokeCallbacks(new_status);
+  }
+};
+
+bool IsConnectivityChange(const Status& status) {
+  return status.code() == FirestoreErrorCode::Unavailable;
+}
+
+class ConnectivityObserver : public GrpcStreamObserver {
+ public:
+  void OnStreamStart() override {
+  }
+  void OnStreamRead(const grpc::ByteBuffer& message) override {
+  }
+  void OnStreamFinish(const util::Status& status) override {
+    if (IsConnectivityChange(status)) {
+      ++connectivity_change_count_;
+    }
+  }
+
+  int connectivity_change_count() const {
+    return connectivity_change_count_;
+  }
+  int connectivity_change_count_ = 0;
+};
+
+}  // namespace
+
+class GrpcConnectionTest : public testing::Test {
+ public:
+  GrpcConnectionTest() {
+    auto connectivity_monitor_owning =
+        absl::make_unique<FakeConnectivityMonitor>(&tester->worker_queue());
```

Comment:
I can't understand what this is doing.

It seems like at the point this runs, tester hasn't been assigned yet, so it would be null. Postfix expressions have precedence over unary operations, so this seems to be invoking worker_queue on a null pointer and then taking its address ... which shouldn't work?

Then separately, we keep a non-owning pointer to the connectivity monitor around to circumvent the fact that the monitor isn't part of the tester's interface.

This all would be much, much simpler and clearer if these cooperating objects didn't have ownership mixed in.

In the production system it's easy to make the FirestoreClient just own the lifetime of everything that needs to participate together. Everything else just gets non-owning pointers to the participants and doesn't have to worry about ownership/lifetime.

That makes the tests easier to write too: just make each of these things members and hand out pointers to the concerned parties.